### PR TITLE
ARROW-7277: [Java] [Doc] Add discussion about vector lifecycle

### PR DESCRIPTION
**Pull request by liyafan82**
As discussed in https://issues.apache.org/jira/browse/ARROW-7254?focusedCommentId=16983284&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16983284, we need a discussion about the lifecycle of a vector.

Each vector has a lifecycle, and different operations should be performed in particular phases of the lifecycle. If we violate this, some unexpected results may be produced. This may cause some confusion for Arrow users. So we want to add a new section to the prose document, to make it clear and explicit.

### CODE REVIEW

**Comment by emkornfield** [comment]
@liyafan82 sorry, I've been backlogged will try to review your open PRs over the next week or so (unless others get to them sooner CC @praveenbingo @pravindra @BryanCutler)

**Comment by liyafan82** [comment]
> @liyafan82 sorry, I've been backlogged will try to review your open PRs over the next week or so (unless others get to them sooner CC @praveenbingo @pravindra @BryanCutler)

@emkornfield Please do not be sorry. You have already given me so much help and guidance. I really appreciate all your effort and professional spirit. So please don't worry, just take your time. 

I think it is late in your location. I also notice that you often work late into night reviewing our PR. So I also hope you can take good care of your health and have some good rest :-)

File: docs/source/java/vector.rst [hunk 1]
```diff
@@ -31,7 +31,87 @@ Table with non-intuitive names (BigInt = 64 bit integer, etc).
...
+    vector.set(/*index*/5, /*value*/25);
+
+4. **Set value count**: for this step, we set the value count of the vector by calling the
+``setValueCount(int)`` method:
+
+.. code-block:: Java
+
+    vector.setValueCount(10);
+
+After this step, the vector enters an immutable state. In other words, we should no longer mutate it.
```
Comment (jacques-n) [inline docs/source/java/vector.rst]: This is true unless we allocate again. Vectors are designed to be reused and thus this allocate, mutate, set valuecount, clear, allocate, mutate pattern will be done repeatedly in most cases.

File: docs/source/java/vector.rst [hunk 2]
```diff
@@ -41,50 +121,56 @@ Note that the current implementation doesn't enforce the rule that Arrow objects
...
-   vector.set(4, 5);
-   vector.set(5, 6);
-   vector.set(6, 7);
-   vector.set(7, 8);
-   vector.setValueCount(8); // this will finalizes the vector by convention.
+vector of the range 0 to 7 where the element that should hold the fourth value is nulled
+
+.. code-block:: Java
+
+    BufferAllocator allocator = new RootAllocator(Long.MAX_VALUE);
```
Comment (jacques-n) [inline docs/source/java/vector.rst]: I suggest you cover here or somewhere that most objects in Arrow implement AutoCloseable and in your demos you use try-with-resources to make that obvious. For example, it should be done here with both allocator and vector.

File: docs/source/java/vector.rst (hunk 1 above)
Comment (liyafan82) [reply docs/source/java/vector.rst]: Thanks for your comments.
I have added a note for this.

File: docs/source/java/vector.rst (hunk 2 above)
Comment (liyafan82) [reply docs/source/java/vector.rst]: Good point. Thanks. 
I have added revised the code snippet, and added a note by the end of the first section.

**Comment by emkornfield** [comment]
Going to merge this.  Thanks.

**Pull request closed without merging**